project-metrics/metrics_service/metrics/base.py

```diff
@@ -0,0 +1,99 @@
+"""Base interface for metrics to implement."""
+
+import abc
+from typing import Sequence, Optional, Text
+import stringcase
+
+import models
+
+
+class Metric(object):
+  """Abstract base class for health metrics."""
+
+  __metaclass__ = abc.ABCMeta
+
+  def __init__(self):
+    self._result = None
+
+  def __str__(self) -> Text:
+    return '%s: %s' % (self.label, self.formatted_result)
+
+  @property
+  def name(self):
+    """Metric name to use as DB identifier when storing results."""
+    return self.__class__.__name__
+
+  @property
+  def label(self) -> Text:
+    """Label for the metric (ex. TestCoverageMetric -> "Test Coverage")."""
+    return stringcase.titlecase(
+        self.name[:-6] if self.name.endswith('Metric') else self.name)
+
+  @property
+  def formatted_result(self) -> Text:
+    """The formatted result value (ex. 3w, 80%, 100PRs)."""
+    return self._format_value(self._result.value) if self._result else '?'
+
+  @property
+  def result(self) -> Optional[models.MetricResult]:
+    """The result of a computation of the metric."""
+    if self._result is None:
+      self._result = self._fetch_result()
```

Comment:
This might cause unexpected behaviour down the line - if this is supposed to show a large table of values and a bunch of them are missing, that will cause a lot of background network or database activity before showing anything, right?

Reply:
This ends up being replaced in the next PR, but for now I'll drop this part.